drivers/modbus_rtu/modbus_rtu.c

```diff
@@ -0,0 +1,469 @@
+      break;
+    case MB_FC_WRITE_REGISTERS:
+      // (id + func + addr + count + size) + data + crc
+      size = 7 + modbus->buffer[BYTE_CNT] + 2;
+      if (wait_bytes(modbus, size)) {
+        err = MB_ER_TIMEOUT;
+        goto exit;
+      }
+      mutex_lock(&(modbus->mutex_buffer));
+      if (calcCRC(modbus->buffer, size) != 0) {
+        mutex_unlock(&(modbus->mutex_buffer));
+        err = MB_ER_CRC;
+        goto exit;
+      }
+      read_address(modbus);
+      read_count(modbus);
+      memcpy(modbus->msg->data, modbus->buffer + 7, modbus->buffer[BYTE_CNT]);
+      mutex_unlock(&(modbus->mutex_buffer));
+      break;
+    default:
+      err = MB_ER_ILLEGAL_FUNCTION;
+      goto exit;
+      break;
+    }
+
+    goto exit;
+  }
+exit:
+  modbus->size_buffer = 0;
+  modbus->msg = NULL;
+  return err;
+}
+
+int modbus_rtu_send_response(modbus_rtu_t *modbus, modbus_rtu_message_t *message) {
+  uint8_t size;
+  div_t d;
+  modbus->msg = message;
+  mutex_lock(&(modbus->mutex_buffer));
+  modbus->buffer[ID] = modbus->msg->id;
+  modbus->buffer[FUNC] = modbus->msg->func;
+  switch (modbus->msg->func) {
+  case MB_FC_READ_COILS:
+  case MB_FC_READ_DISCRETE_INPUT:
+    d = div(modbus->msg->count, 8);
+    size = d.quot;
+    if (d.rem) {
+      size += 1;
+    }
+    modbus->buffer[2] = size;
+    modbus->size_buffer = 3 + size;
+    copy_bits(modbus->buffer + 3, 0, (uint8_t *)modbus->msg->data, modbus->msg->addr, modbus->msg->count);
+    break;
+  case MB_FC_WRITE_COIL:
+  case MB_FC_WRITE_REGISTER:
+  case MB_FC_WRITE_COILS:
+  case MB_FC_WRITE_REGISTERS:
+    // don't change any data in buffer, send this back as is
+    // size_buffer change for calculate CRC
+    // todo: calculate CRC for MB_FC_WRITE_COIL no need, need reduce this
+    // id + func + addr + val
+    modbus->size_buffer = 6;
+    break;
+  case MB_FC_READ_REGISTERS:
+  case MB_FC_READ_INPUT_REGISTER:
+    size = modbus->msg->count * 2;
+    modbus->buffer[2] = size; // calc size for send
+    modbus->size_buffer = 3 + size;
+    memcpy(modbus->buffer + 3, modbus->msg->data, size);
+    break;
+  default:
+    mutex_unlock(&(modbus->mutex_buffer));
+    return MB_ER_ILLEGAL_FUNCTION;
+    break;
+  }
+  send(modbus);
+  modbus->size_buffer = 0;
+  mutex_unlock(&(modbus->mutex_buffer));
+  return MB_ER_NONE;
+}
+
+static inline void send(modbus_rtu_t *modbus) {
+  uint16_t crc = calcCRC(modbus->buffer, modbus->size_buffer);
+  modbus->buffer[modbus->size_buffer] = lowByte(crc);
+  modbus->size_buffer++;
+  modbus->buffer[modbus->size_buffer] = highByte(crc);
+  modbus->size_buffer++;
+  enamdle_trasmit(modbus);
+  uart_write(modbus->uart, modbus->buffer, modbus->size_buffer);
+  disamdle_trasmit(modbus);
+}
+
+static inline void enamdle_trasmit(modbus_rtu_t *modbus) {
+  if (modbus->pin_rts != GPIO_UNDEF) {
+    gpio_write(modbus->pin_rts, modbus->pin_rts_enable);
+  }
+}
+
+static inline void disamdle_trasmit(modbus_rtu_t *modbus) {
```

Comment:
disable?

Reply:
`enamdle_trasmit`  set `pin_rts` to value `pin_rts_enable`, and `disamdle_trasmit` to value `!enamdle_trasmit`.

Reply:
i'm don't know how name this better.